api/v1alpha1/externaldns_types.go

```diff
@@ -18,41 +18,412 @@ package v1alpha1
+	// An entry that excludes a domain that isn't a subdomain of some
+	// included domain will be ignored.
+	//
+	// An empty list of domains means ExternalDNS will create
+	// DNS records for any included source resource regardless
+	// of the resource's desired hostname.
+	//
+	// +optional
+	Domains []ExternalDNSDomain `json:"domains,omitempty"`
+
+	// Provider refers to the DNS provider that ExternalDNS
+	// should publish records to. Note that each ExternalDNS
+	// is tied to a single provider.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	Provider ExternalDNSProvider `json:"provider"`
 
-	// Foo is an example field of ExternalDNS. Edit externaldns_types.go to remove/update
-	Foo string `json:"foo,omitempty"`
+	// Source describes which source resource
+	// ExternalDNS will be configured to create
+	// DNS records for.
+	//
+	// Multiple ExternalDNS CRs must be
+	// created if multiple ExternalDNS source resources
+	// are desired.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	Source ExternalDNSSource `json:"source"`
+
+	// Zones describes which DNS Zone IDs
+	// ExternalDNS should publish records to.
+	//
+	// +optional
+	Zones []string `json:"zones,omitempty"`
 }
 
-// ExternalDNSStatus defines the observed state of ExternalDNS
+// ExternalDNSDomain describes how sets of included
+// or excluded domains are to be constructed.
+type ExternalDNSDomain struct {
+	ExternalDNSDomainUnion `json:",inline"`
+
+	// FilterType marks the Name or Pattern field
+	// as an included or excluded set of domains.
+	//
+	// Note that excluded domains that
+	// have not been included will be
+	// ignored.
+	//
+	// This field accepts the following values:
+	//
+	//  "Include": Include the domain set specified
+	//  by name or pattern.
+	//
+	//  "Exclude": Exclude the domain set specified
+	//  by name or pattern.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	FilterType ExternalDNSFilterType `json:"filterType"`
+}
+
+// ExternalDNSDomainUnion describes optional fields of an External domain
+// that should be captured.
+// +union
+type ExternalDNSDomainUnion struct {
+	// MatchType specifies the type of match to be performed
+	// by ExternalDNS when determining whether or not to publish DNS
+	// records for a given source resource based on the resource's
+	// requested hostname.
+	//
+	// This field accepts the following values:
+	//
+	//  "Exact": Explicitly match the full domain string
+	//   specified via the name field, including any subdomains
+	//   of the given name.
+	//
+	//  "Pattern": Match potential domains against
+	//  the provided regular expression pattern string.
+	//
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	// +required
+	MatchType DomainMatchType `json:"matchType"`
+
+	// Name is a string representing a single domain
+	// value.
+	//
+	// e.g. my-app.my-cluster-domain.com
+	//
+	// +optional
+	Name *string `json:"names,omitempty"`
+
+	// Pattern is a regular expression used to
```

Comment:
you need to specify the regex language. They aren't all the same.

Reply:
Right. I've spiced this comment up. Thanks!